paper.md

```diff
@@ -38,89 +38,140 @@ affiliations:
+
+To overcome these two issues, we designed a new type of integration question described in the following subsection.
 
 ## DSW Widget Integrations
 
-*TODO: Jan - describe the widget integration question*
+Although the integration question was a good start, we needed some improvements. The main issue was that the standard integration question in DSW connects to a public API of an external service to find the results. However, in DAISY, users need to log in first to access their projects. So the standard solution was not suitable unless all of the projects in DAISY were public, which would not be desirable.
+
+Therefore, we came up with the idea of an integration widget. Instead of a search field in DSW, we can open a popup window with a web page implemented on the connected service side, in this case, DAISY. The implementation of the widget itself is the responsibility of the connected service. That means that any necessary functionality, such as login, can be implemented.
+
+We implemented the [DSW Integration Widget SDK](https://github.com/ds-wizard/dsw-integration-widget-sdk), a JavaScript library that is a wrapper for sending the selected options back to DSW, and corresponding supporting code on the DSW side. The library and DSW uses JavaScript `Window.postMessage()` under the hood for the communication between DSW and the popup window with the widget. Besides sending the answers back to DSW, the library also checks whether the widget was invoked from the allowed DSW instance to avoid leaking information somewhere unwanted.
+
+Once the question is answered using the widget, it works in the same way as the standard integration question regarding how it is saved. We save the name of the project retrieved from DAISY and the url to the project so it can be used for further processing.
 
-*TODO: Jan - screenshot of the DSW side, code snippet*
+![Integration widget question in DSW -- empty and with selected answer from DAISY](figures/dsw-daisy-question.png)
 
 ## Widget for DAISY Projects
 
-*TODO: Vilém (and Jan) - describe implementation on DAISY side*
+The widget implemented on DAISY's side allows users to select one of existing projects and send them back as an answer to DSW. When the widget is opened in a popup window, user is asked to log into DAISY. As the initial page is actual DAISY login form, user needs to use the DAISY login credentials. Once logged in, user is redirected to `/integrations/dsw/list-projects` DAISY endpoint presenting a list of projects he has access to. The list contains only records for which the user is appointed to as local custodian or personnel. The internal DAISY project ID, project's acronym and full title is presented. When a project of interest is clicked on, the popup window is closed, and the information is set in DSW.
```

Comment:
The widget implemented on DAISY's side allows users to select one of the existing projects and send them back as an answer to DSW. When the widget is opened in a popup window, users are asked to log into DAISY. As the initial page is the actual DAISY login form, the DAISY login credentials are required. Once logged in, users are redirected to the `/integrations/dsw/list-projects` DAISY endpoint presenting a list of projects they have access to. The list contains only records for which users are appointed to as local custodians or personnel. The internal DAISY project ID, project's acronym and full title is presented. When a project of interest is clicked on, the popup window is closed, and the information is sent to DSW.

Reply:
```suggestion
The widget implemented on DAISY's side allows users to select one of the existing projects and send them back as an answer to DSW. When the widget is opened in a popup window, users are asked to log into DAISY. As the initial page is the actual DAISY login form, the DAISY login credentials are required. Once logged in, users are redirected to the `/integrations/dsw/list-projects` DAISY endpoint presenting a list of projects they have access to. The list contains only records for which users are appointed to as local custodians or personnel. The internal DAISY project ID, project's acronym and full title is presented. When a project of interest is clicked on, the popup window is closed, and the information is sent to DSW.
```